src/index.ts

```diff
@@ -315,413 +317,183 @@ export async function onFormSubmit(
-          } = groupDocsByTargetDocPath(highPriorityDstPathLogicDocsMap, docPath);
-          await distribute(highPriorityDocsByDocPath);
-          await distribute(highPriorityOtherDocsByDocPath);
-
-          if (page === 0) {
-            await formRef.update({"@status": "finished"});
-          }
+      if (runStatus.status === "cancel-then-retry") {
+        await formRef.update({"@status": "cancelled", "@messages": "cancel-then-retry received " +
+              "from business logic"});
+        return;
+      }
 
-          console.info(`Consolidating and Distributing Normal Priority Logic Results: ${normalPriorityDocs.length}`);
-          const normalPriorityDstPathLogicDocsMap: Map<string, LogicResultDoc[]> =
-              await expandConsolidateAndGroupByDstPath(normalPriorityDocs);
-          const {
-            docsByDocPath: normalPriorityDocsByDocPath,
-            otherDocsByDocPath: normalPriorityOtherDocsByDocPath,
-          } = groupDocsByTargetDocPath(normalPriorityDstPathLogicDocsMap, docPath);
-          await distribute(normalPriorityDocsByDocPath);
-          await distributeLater(normalPriorityOtherDocsByDocPath);
-
-          console.info(`Consolidating and Distributing Low Priority Logic Results: ${lowPriorityDocs.length}`);
-          const lowPriorityDstPathLogicDocsMap: Map<string, LogicResultDoc[]> =
-              await expandConsolidateAndGroupByDstPath(lowPriorityDocs);
-          const {
-            docsByDocPath: lowPriorityDocsByDocPath,
-            otherDocsByDocPath: lowPriorityOtherDocsByDocPath,
-          } = groupDocsByTargetDocPath(lowPriorityDstPathLogicDocsMap, docPath);
-          await distributeLater(lowPriorityDocsByDocPath);
-          await distributeLater(lowPriorityOtherDocsByDocPath);
+      function updateErrorMessage() {
+        const errorLogicResults = runStatus.logicResults.filter((result) => result.status === "error");
+        if (errorLogicResults.length > 0) {
+          errorMessage = errorMessage + errorLogicResults.map((result) => result.message).join("\n");
         }
-        await distributeNonTransactionalLogicResults();
       }
-    );
-    const businessLogicEnd = performance.now();
-    const businessLogicLogicResult: LogicResult = {
-      name: "runBusinessLogics",
-      status: "finished",
-      documents: [],
-      execTime: businessLogicEnd - businessLogicStart,
-    };
-    logicResults.push(businessLogicLogicResult);
+      updateErrorMessage();
 
-    if (runStatus === "cancel-then-retry") {
-      await formRef.update({"@status": "cancelled", "@messages": "cancel-then-retry received " +
-              "from business logic"});
+      const distributeLogicResultsStart = performance.now();
+      await distributeFnTransactional(txn, runStatus.logicResults);
+      const distributeLogicResultsEnd = performance.now();
+      const distributeLogicResultPerf: LogicResult = {
+        name: "distributeLogicResults",
```

Comment:
let's rename this to distributeTransactionalLogicResults.  Let's also measure performrnace of distributeNonTransactionalLogicResults